### ISSUE TITLE
gzip support?

### ISSUE THREAD

**Issue by makc** [pwnall/node-xhr2#17]
I was getting 404 for some file before I did this
```
    xhr.open(method, url, true)
    if XMLHttpRequest.prototype._restrictedHeaders
        delete XMLHttpRequest.prototype._restrictedHeaders['accept-encoding']
        xhr.setRequestHeader('Accept-Encoding', 'gzip, deflate')
```
Now I don't get 404, but the data is obviously garbage. Is there easy way to get unzipped content in this case?

**Comment by pwnall** [comment]
The server must have a weird configuration.

I'd look into zlib, it is built into node.js (at least for 0.10 and 0.12).
https://nodejs.org/docs/v0.10.0/api/zlib.html

This sounds like an useful thing to support out of the box. Sorry I don't have time to work on it right now! We can leave the issue open for when I have time to get to it.

**Comment by makc** [comment]
yes, server does not serve the file non-zipped for some reason. I ended up using [https://github.com/request/request](request)

**Comment by jimmyangel** [comment]
Do you have any plans to add gzip support to this library? Thank you.

**Comment by destructive-dragon** [comment]
Angular's Server Side rendering depends on this project and requires GZIP support :(

**Comment by pwnall** [comment]
For Angular folks -- I'm at Google now, so I could justify spending some time on this if it's helpful to Angular. Please reach out to me if you'd like to discuss.

**Comment by benjamingr** [comment]
I patched this into my own implementation with the following diff at Testim: 

```diff
             } else {
                 buffer = this._concatBuffers(this._responseParts);
             }
+            // temporary workaround for transport encoding parsing
+            if (this._responseHeaders['content-encoding'] && this._responseHeaders['content-encoding'] === 'gzip') {
+                buffer = require('zlib').gunzipSync(buffer);
+            }
+            console.log('response parse', this._responseHeaders);
             this._responseParts = null;
             switch (this.responseType) {
                 case 'text':
```

Feel free to use this (it's _very basic_ and mostly incorrect since it doesn't stream the zipped data), a more correct implementation would reuse the logic for this from superagent or request. 

It worked for me, feel free to use this.

**Comment by mahpour** [comment]
> I patched this into my own implementation with the following diff at Testim:
> 
> ```diff
>              } else {
>                  buffer = this._concatBuffers(this._responseParts);
>              }
> +            // temporary workaround for transport encoding parsing
> +            if (this._responseHeaders['content-encoding'] && this._responseHeaders['content-encoding'] === 'gzip') {
> +                buffer = require('zlib').gunzipSync(buffer);
> +            }
> +            console.log('response parse', this._responseHeaders);
>              this._responseParts = null;
>              switch (this.responseType) {
>                  case 'text':
> ```
> 
> Feel free to use this (it's _very basic_ and mostly incorrect since it doesn't stream the zipped data), a more correct implementation would reuse the logic for this from superagent or request.
> 
> It worked for me, feel free to use this.

Can you please explain where/how this should be applied?

**Comment by benjamingr** [comment]
This is a diff file, you can typically apply it with the `patch` command http://man7.org/linux/man-pages/man1/patch.1.html

**Comment by mahpour** [comment]
> This is a diff file, you can typically apply it with the `patch` command http://man7.org/linux/man-pages/man1/patch.1.html

I understand. but which file?

**Comment by mahpour** [comment]
I figured it out. I used the script patching to apply the patch after npm install. 
Thanks to John Liu from: 
http://johnliu.net/blog/2018/12/hot-patching-our-dependencies-with-patch-package

**Comment by RomanChygryn** [comment]
Is this still not supported by Angular Universal?
Logs from ssr throws mi this error Refused to set unsafe header "Accept-Encoding"
I added "Accept-Encoding" header to all requests to enable gzip.

I need to know if i understand correctly: Angular Universal use xhr2 for all requests, and xhr2 not support gzip, thats what my ssr refuse "Accept-Encoding" header, i'm right?

**Comment by amakhrov** [comment]
`"Accept-Encoding"` is not supposed to be set manually at all. Instead, user agent is expected to set it automatically for all outgoing requests, based on what this user agent really supports.

Now, the real question is - does `xhr2` support decoding gzipped responses? If yes - it should set this header automatically. Otherwise - curious to know if there are any plans to start supporting it.

**Comment by mahpour** [comment]
I checked xhr2 used in angular 9 universal and it still doesn’t support gzip decompression. I had to create a new patch file and use it in our build process.
